connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/ExecutePlanResponseReattachableIterator.scala

```diff
@@ -297,6 +297,6 @@ class ExecutePlanResponseReattachableIterator(
   /**
    * Retries the given function with exponential backoff according to the client's retryPolicy.
    */
-  private def retry[T](fn: => T, currentRetryNum: Int = 0): T =
-    GrpcRetryHandler.retry(retryPolicy)(fn, currentRetryNum)
+  private def retry[T](fn: => T): T =
+    GrpcRetryHandler.retry(retryPolicy)(fn)
```

Comment:
Wow, this is pretty complicated design in there. Can the code there be changed to just use ```GrpcRetryHandler.retry```?

Something like
```
GrpcRetryHandler.retry(retryPolicy)(() =>     new StreamObserver[proto.ReleaseExecuteResponse] {
      override def onNext(v: proto.ReleaseExecuteResponse): Unit = {}
      override def onCompleted(): Unit = {}
      override def onError(t: Throwable): Unit = throw t
})

Reply:
If it hopped to another grpc thread, I assume all **subsequent** retries would still be executed on that thread?

Can we do one iterator of the current code and then switch to GrpcRetryHandler.retry instead of further tail recursion?

Reply:
Discussed offline,
```
      override def onError(t: Throwable): Unit = {
        var firstTry = true
        try {
          retry {
            if (firstTry) {
              firstTry = false
              throw t // we already failed once, handle first retry
            } else {
              rawBlockingStub.releaseExecute(requestForRetry)
            }
          }
        } catch {
          case NonFatal(e) =>
            logWarning(s"ReleaseExecute failed with exception: $e.")
        }
```
should work and makes it better because it reuses the retry logic. After the first failure, we are already on the GRPC async execution thread, and can iterate with a retry loop there instead of recursing. This also brings it closer to the python implementation.

Reply:
I spinned the refactoring of createRetryingReleaseExecuteResponseObserer to https://github.com/apache/spark/pull/42438

Reply:
Thank you very much @juliuszsompolski!